libtock/alarm_timer.c

```diff
@@ -106,6 +131,84 @@ int alarm_at(uint32_t reference, uint32_t dt, subscribe_upcall cb, void* ud, ala
   return RETURNCODE_SUCCESS;
 }
 
+// Holds original callback and user data. Also keeps bookkeeping for the
+// intermediate alarm overflows.
+typedef struct overflow_ud {
+  // How many times the alarm has left to overflow before reaching target time
+  int overflows_left;
+  // Number of remaining ticks after last full alarm overflow
+  int remainder_ticks;
+  // Original user data passed to `timer_in`
+  void* original_ud;
+  // Original callback passed to `timer_in`
+  subscribe_upcall* original_cb;
+  alarm_t* alarm;
+} overflow_ud_t;
+
+// The intermediate callback that handles overflows of alarm. This is used by
+// `timer_in` to handle cases where `alarm_at` would normally be unable to
+// support the whole timer length. `alarm_at` can only keep track of up to 2^32
+// ticks.
+static void overflow_callback(int                          last_timer_fire_time,
+                              __attribute__ ((unused)) int unused1,
+                              __attribute__ ((unused)) int unused2,
+                              void*                        overflow_ud) {
+  overflow_ud_t* ud = (overflow_ud_t*)overflow_ud;
+
+  if (ud->overflows_left == 0) {
+    // no overflows left, schedule last alarm with original callback
+    alarm_at(
+      last_timer_fire_time,
+      ud->remainder_ticks,
+      ud->original_cb,
+      ud->original_ud,
+      ud->alarm
+      );
+  } else {
+    // schedule next intermediate alarm that will overflow
+    ud->overflows_left--;
+
+    alarm_at(
+      last_timer_fire_time,
+      MAX_TICKS,
+      (subscribe_upcall*) overflow_callback,
+      (void*) ud,
+      ud->alarm
+      );
+  }
+}
+
+// Timer implementation
+int timer_in(uint32_t dt_ms, subscribe_upcall cb, void* ud, tock_timer_t *tock_timer) {
+  alarm_t* alarm = &(tock_timer->alarm);
+  uint32_t now;
+  int ret = alarm_internal_read(&now);
+  if (ret != RETURNCODE_SUCCESS) return ret;
+
+  // If `dt_ms` is longer than the time that an alarm can count up to, then `timer_in` will
+  // schedule multiple alarms to reach the full length. We calculate the number of full overflows
+  // and the remainder ticks to reach the target length of time. The overflows use the
+  // `overflow_callback` for each intermediate overflow.
+  uint32_t max_ticks_in_ms = ticks_to_ms(MAX_TICKS);
+  if (dt_ms > max_ticks_in_ms) {
+    overflow_ud_t* tmp_ud = malloc(sizeof(overflow_ud_t));
```

Comment:
I don't see a `free` in this patch.

In general, we want to move away from libtock holding internal state. Can this just be part of the timer object?

Reply:
> I don't see a free in this patch.

Sorry! Can't believe I missed the oldest trick in the book.

> Can this just be part of the timer object?

Good point. This could be added to the timer object. This will require changing how the timer length is stored from `uint32_t interval;` to `int overflows_left;` plus `int remainder_ticks;`.

Reply:
Resolved in https://github.com/tock/libtock-c/pull/395/commits/f0b0b4691fad3a091da2768953feb9783234ddca

We no longer have any frees/mallocs. Just modified the fields of `tock_timer_t` to support the overflow handling.

Reply:
We verified that our test with `timer_in` still works as intended and that `timer_every` still works by running `examples/tests/multi_alarm_test`